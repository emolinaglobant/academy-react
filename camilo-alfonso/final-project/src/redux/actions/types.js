export const ActionTypes = {
  SET_ROCKETS: "SET_ROCKETS",
  SELECT_ROCKET: "SELECT_ROCKET",
  REMOVE_ROCKET: "REMOVE_ROCKET",
  SET_LAUNCHES: "SET_LAUNCHES",
  SET_LAUNCH: "SET_LAUNCH",
  SELECT_LAUNCH: "SELECT_LAUNCH",
  SET_ASTRONAUTS: "SET_ASTRONAUTS",
  SELECT_ASTRONAUT: "SELECT_ASTRONAUT",
  SET_ROADSTER: "SET_ROADSTER",
};
