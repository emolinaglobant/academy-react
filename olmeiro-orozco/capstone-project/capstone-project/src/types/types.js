export const types = {
  login: '[Auth] Login',
  logout: '[Auth] Logout',

  uiSetError: '[UI] Set Error',
  uiRemoveError: '[UI] Remove Error',
  uiStartLoading: '[UI] Start loading',
  uiFinishLoading: '[UI] Finish loading',

  startRound: '[Game] Start Round',
  gameStoreQuestions: '[Game] Store Questions',
  gameIncreasePrize: '[Game] Increase Prize',
  gameAddRound: '[Game] Add Round',
  gameChangeRound: '[Game] Change Round',
  gameReset: '[Game] Reset',
}
