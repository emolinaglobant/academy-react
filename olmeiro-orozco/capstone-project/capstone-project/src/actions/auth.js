import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from "firebase/auth";
import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import Swal from "sweetalert2";

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());

    const auth = getAuth();
    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));

        dispatch(finishLoading());
      })
      .catch((e) => {
        dispatch(finishLoading());
        Swal.fire({
          icon: 'error',
          title: 'Oops...',
          text: 'Something went wrong!',
          footer: 'Please confirm your Email and Password.'
        })
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    const auth = getAuth();
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });

        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => {
        console.log(error);
        Swal.fire("Error", error.message, "error");
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName));
    });
  };
};

export const login = (uid, displayName) => {
  return {
    type: types.login,
    payload: {
      uid,
      displayName,
    },
  };
};

export const startLogout = () => {
  return async (dispath) => {
    const auth = getAuth();
    await signOut(auth);

    dispath(logout());
  };
};

export const logout = () => {
  return {
    type: types.logout,
  };
};
