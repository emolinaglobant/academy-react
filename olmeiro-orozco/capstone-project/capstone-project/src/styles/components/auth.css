.auth__main {
  align-items: center;
  background-color: var(--BlueHaze);
  background-image: linear-gradient(rgba(0, 0, 255, 0.5), rgba(255, 255, 0, 0.5)), url("../../../public/assets/trivia2.jpg");
  display: flex;
  height: 100vh;
  justify-content: center;
  margin: 0px;
  width: 100vw;
}

.auth__box-container {
  background-color: var(--white);
  border-radius: 2px;
  box-shadow: 0px 3px var(--CopperRose);
  padding: 20px;
  width: 250px;
}

.auth__title {
  color: var(--darkgray);
  margin-bottom: 10px;
}

.auth__input {
  border-bottom: 1px solid var(--darkgray);
  border: 0px;
  color: var(--darkgray);
  font-size: 16px;
  height: 20px;
  margin-bottom: 20px;
  transition: border-bottom .3s ease-out;
  width: 100%;
}

.auth__input:focus {
  outline: none;
  border-bottom: 1px solid var(--primary);
}

.auth__google {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding-bottom: 20px;
  padding-top: 20px;
  width: 100%;
}

.auth__alert-error {
  background-color: var(--red);
  border-radius: 5px;
  color: var(--secondary);
  display: flex;
  justify-content: center;
  margin-bottom: 10px;
  padding: 5px;
}