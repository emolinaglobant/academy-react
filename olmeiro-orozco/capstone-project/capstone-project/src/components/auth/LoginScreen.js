import React from "react";
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { startGoogleLogin, startLoginEmailPassword } from "../../actions/auth";

import { useForm } from "../../hooks/useForm";

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const { loading } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: "",
    password: "",
  });

  const { email, password } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginEmailPassword(email, password));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <section>
      <h3 className="auth__title">Login</h3>
      <form onSubmit={handleLogin}>
        <input
          autoComplete="off"
          className="auth__input"
          name="email"
          onChange={handleInputChange}
          placeholder="email"
          type="text"
          value={email}
        />
        <input
          className="auth__input"
          name="password"
          onChange={handleInputChange}
          placeholder="password"
          type={"password"}
          value={password}
        />

        <button className="btn btn__primary" disabled={loading} type="submit">
          Login
        </button>

        <section className="auth__google">
          <p>Login with google</p>

          <section className="google__btn" onClick={handleGoogleLogin}>
            <section className="google__icon-wrapper">
              <img
                alt="google button"
                className="google__icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
              />
            </section>
            <p className="btn__text">
              <b>Sign in with google</b>
            </p>
          </section>
        </section>
        <Link to="/auth/register" className="link">
          Create new account
        </Link>
      </form>
    </section>
  );
};
