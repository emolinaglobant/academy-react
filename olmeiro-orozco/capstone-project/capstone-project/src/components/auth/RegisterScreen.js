import React from "react";
import { Link } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import validator from "validator";
import { removeError, setError } from "../../actions/ui";
import { startRegisterWithEmailPasswordName } from "../../actions/auth";
import { useSelector, useDispatch } from "react-redux";

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { msgError } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    name: "",
    email: "",
    password: "",
    password2: "",
  });

  const { name, email, password, password2 } = formValues;

  const handleRegister = (e) => {
    e.preventDefault();

    if (isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(email, password, name));
    }
  };

  const isFormValid = () => {
    if (validator.isEmpty(name)) {
      dispatch(setError("name is required"));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError("Email is required"));
      return false;
    } else if (password !== password2 || password.length < 5) {
      dispatch(
        setError(
          "passwor should be at least six characters and match each other"
        )
      );
      return false;
    }
    dispatch(removeError());
    return true;
  };

  return (
    <section>
      {msgError && <section className="auth__alert-error">{msgError}</section>}

      <h3 className="auth__title">Register</h3>

      <form onSubmit={handleRegister}>
        <input
          className="auth__input"
          type="text"
          placeholder="name"
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          value={email}
          onChange={handleInputChange}
          type="email"
          placeholder="email"
          name="email"
          autoComplete="off"
        />
        <input
          className="auth__input"
          type={"password"}
          placeholder="password"
          name="password"
          value={password}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type={"password"}
          placeholder="confirm password"
          name="password2"
          value={password2}
          onChange={handleInputChange}
        />

        <button type="submit" className="btn btn__primary">
          Register
        </button>

        <Link to="/auth/login" className="link">
          Already registered?
        </Link>
      </form>
    </section>
  );
};
