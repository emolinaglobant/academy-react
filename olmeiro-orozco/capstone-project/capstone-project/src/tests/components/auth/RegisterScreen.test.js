import React from "react";
import thunk from "redux-thunk";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";

import { types } from "../../../types/types";
import { RegisterScreen } from "../../../components/auth/RegisterScreen";

jest.mock("../../../actions/auth", () => ({
  startGoogleLogin: jest.fn(),
  startLoginEmailPassword: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null,
  },
};

let store = mockStore(initState);

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <RegisterScreen />
    </MemoryRouter>
  </Provider>
);

describe("Tests component <RegisterScreeen /> ", () => {
  test("should show it correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should dispatch correct action input name', () => { 
    const nameInput = wrapper.find('input[name="name"]');

    nameInput.simulate('change', {
      target: {
        value: '',
        name:'name'
      }
    });

    wrapper.find('form').simulate('submit', {
      preventDefault(){}
    });

    const actions = store.getActions();

    expect(actions[0]).toEqual({
      type: types.uiSetError,
      payload: 'name is required'
    })
  });
 
  test("should dispatch correct action input email", () => {
    const nameInput = wrapper.find('input[name="name"]');
    const emailInput = wrapper.find('input[name="email"]');

    nameInput.simulate("change", {
      target: {
        value: "Test input",
        name: "name",
      },
    });

    emailInput.simulate("change", {
      target: {
        value: "",
        name: "email",
      },
    });

    wrapper.find("form").simulate("submit", {
      preventDefault() {},
    });

    const actions = store.getActions();
   
    expect(actions[1]).toEqual({
      type: types.uiSetError,
      payload: "Email is required",
    });
  });
 
  test("should dispatch correct action input password not match", () => {
    const nameInput = wrapper.find('input[name="name"]');
    const emailInput = wrapper.find('input[name="email"]');
    const passwordInput = wrapper.find('input[name="password"]');
    const password2Input = wrapper.find('input[name="password2"]');

    nameInput.simulate("change", {
      target: {
        value: "Test input",
        name: "name",
      },
    });

    emailInput.simulate("change", {
      target: {
        value: "test@gmail.com",
        name: "email",
      },
    });

    passwordInput.simulate("change", {
      target: {
        value: "123456",
        name: "password",
      },
    });
  
    password2Input.simulate("change", {
      target: {
        value: "765432",
        name: "password",
      },
    });

    wrapper.find("form").simulate("submit", {
      preventDefault() {},
    });

    const actions = store.getActions();
    expect(actions[2]).toEqual({
      type: types.uiSetError,
      payload: 'passwor should be at least six characters and match each other'
    });
  });

  test('should show error box', () => { 
    const initState = {
      auth: {},
      ui: {
        loading: false,
        msgError: 'Email is not correct',
      },
    };
    
    const store = mockStore(initState);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
    expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError);
   })
});
