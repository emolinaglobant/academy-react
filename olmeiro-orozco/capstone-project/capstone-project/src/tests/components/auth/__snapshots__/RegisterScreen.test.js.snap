// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests component <RegisterScreeen />  should show it correctly 1`] = `
<section>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="email"
      type="email"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn__primary"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>
</section>
`;
